### ISSUE TITLE
Borg repository size discrepancy between info and real disk usage

### ISSUE THREAD

**Issue by bootc** [borgbackup/borg#6106]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Question (though it may be a bug)

#### Your borg version (borg -V).

borg 1.1.9 (1.1.9-2+deb10u1 from Debian)

#### Operating system (distribution) and version.

Debian 10 (buster)

#### Hardware / network configuration, and filesystems used.

One client backing up to one server over SSH. The repository sits on top of XFS.

#### How much data is handled by borg?

10-17TB

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

Effectively, `borg info`.

## Describe the problem you're observing.

I'm seeing a total deduplicated repository size of 10TB at the moment:

```
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
All archives:              439.34 TB            364.21 TB             10.04 TB
                       Unique chunks         Total chunks
Chunk index:                 9125018            312536348
```

However, the disk usage on the borg repository volume and data directory is 17TB:

```
# df -h /srv/borg
Filesystem               Size  Used Avail Use% Mounted on
/dev/mapper/rootVG-borg   20T   17T  3.5T  83% /srv/borg
# du -sh /srv/borg/repo/data
17T	/srv/borg/repo/data
```

Why is the difference so large, and what can be done about it?

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

I have other pairs of machines with discrepancies in their repository size, but nothing anywhere near this extent.

#### Include any warning/errors/backtraces from the system logs

N/A

**Comment by ThomasWaldmann** [comment]
Just to make sure we're not talking about some already fixed problem: can you use some more recent borg, like this:

https://packages.debian.org/search?suite=buster-backports&searchon=names&keywords=borgbackup

After installing that, do a prune or create/delete 1 archive manually and check space usage again.

If that does not help, try borg check on the repo.

**Comment by bootc** [comment]
I've installed the back-ported BorgBackup and we'll see what happens tomorrow after a backup and prune. I'll let you know.

A full borg check takes the best part of 20 hours so I may not be able to run that immediately.

**Comment by bootc** [comment]
A couple more archives have been added, and a couple have been pruned. Unfortunately that has made no difference in the disk usage. I'm now running a `borg check --archives-only` which only takes a couple of hours; hopefully that's enough to make a difference...

**Comment by bootc** [comment]
> I'm now running a `borg check --archives-only` which only takes a couple of hours; hopefully that's enough to make a difference...

This hasn't made any difference.

**Comment by ThomasWaldmann** [comment]
What you could try is to clear the clientside cache. The chunks index has reference counts and if they somehow are wrong (but their correct refcount would be 0), it might not remove these chunks.

```
borg delete --cache-only REPO
borg prune ... REPO
```

It will then rebuild the caches (might take a while and might use quite some space for cached `chunks.archive.d`, see docs).

**Comment by bootc** [comment]
OK I did that, in fact I deleted all the caches I could find for those repositories, then re-ran the prune. Sadly there was no change in disk usage.

I didn't mention it before, not sure if it changes anything, but the client connects to the repository in append-only mode; the prune happens on the server separately.

**Comment by ThomasWaldmann** [comment]
Then any operation on the client won't reduce disk space usage, of course.

So you deleted the cache on the server, where prune is running (and where it is **not** in append-only mode)?

**Comment by bootc** [comment]
> So you deleted the cache on the server, where prune is running (and where it is **not** in append-only mode)?

Indeed. The only think I haven't done so far is run a full `borg check` without `--archives-only`.

**Comment by bootc** [comment]
Just so I'm clear here, what are the next steps? Do I need to organise the full `borg check` on the repo or is there something else we can do or look up to figure out what's going on?

**Comment by ThomasWaldmann** [comment]
... and also check whether a snapshot is keeping disk space allocated (like e.g. a zfs snapshot).

**Comment by thiagomacieira** [comment]
Sorry to relaunch a closed discussion, but I'm seeing a similar issue, using Borg 1.2.1, using a local repository on an ext4 filesystem.

Before:
```
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
All archives:              234.54 GB            183.54 GB             14.97 GB

                       Unique chunks         Total chunks
Chunk index:                  124177              1343759
$ du -sh repos/home 
21G     repos/home
```

Then I did the dance of `check` (without --repair), delete the cache, create an archive, remove the archive, then did `prune`. Prune finished with:
```
Storage quota: 21.70 GB out of 30.00 GB used.                                                                                                                    
------------------------------------------------------------------------------                                                                                   
                       Original size      Compressed size    Deduplicated size
Deleted data:             -121.39 GB            -96.21 GB             -4.14 GB
All archives:              113.37 GB             87.38 GB             10.83 GB

                       Unique chunks         Total chunks
Chunk index:                   98623               564107
------------------------------------------------------------------------------
$ du -sh repos/home 
21G     repos/home
```

So it made the situation worse.

**Comment by ThomasWaldmann** [comment]
Since borg 1.2 you need to run `borg compact` to really free disk space, see the docs and changelog.

**Comment by thiagomacieira** [comment]
Thank you for the response. Indeed that helped.
```
$ borg compact repos/home          
segment 944 not found, but listed in compaction data
segment 948 not found, but listed in compaction data
segment 952 not found, but listed in compaction data
segment 956 not found, but listed in compaction data
segment 960 not found, but listed in compaction data
segment 964 not found, but listed in compaction data
segment 968 not found, but listed in compaction data
segment 972 not found, but listed in compaction data
segment 976 not found, but listed in compaction data
segment 980 not found, but listed in compaction data
segment 984 not found, but listed in compaction data
segment 988 not found, but listed in compaction data
segment 992 not found, but listed in compaction data
segment 996 not found, but listed in compaction data
segment 1000 not found, but listed in compaction data
segment 1004 not found, but listed in compaction data
segment 1008 not found, but listed in compaction data
segment 1012 not found, but listed in compaction data
segment 1016 not found, but listed in compaction data
segment 1020 not found, but listed in compaction data
segment 1024 not found, but listed in compaction data
segment 1028 not found, but listed in compaction data
segment 1032 not found, but listed in compaction data
$ borg info repos/home             
[...]
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
All archives:              113.37 GB             87.38 GB             10.83 GB

                       Unique chunks         Total chunks
Chunk index:                   98623               564107
$ du -sh repos/home
11G     repos/home
```

**Comment by ThomasWaldmann** [comment]
Looks a bit suspicious, maybe run `borg check --repair REPO`.

**Comment by thiagomacieira** [comment]
Will do, thanks.

**Comment by thiagomacieira** [comment]
```
Storage quota: 10.87 GB out of 20.00 GB used.
Archive consistency check complete, no problems found.
```

I think those refer to the 17-byte files that my script deletes.

**Comment by kmille** [comment]
TLDR: what helped me was: 
`borg compact --verbose --progress server:backup-dir`

**Comment by codethief** [comment]
I can't believe I wasted several days on debugging why my repository was increasing in size every day despite pruning it, only to find out this is a… feature. Please, next time you introduce such a breaking change in Borg, have the `borg` command output a notification until the user explicitly confirms it. And please have `borg info` output disk usage and unused/non-compacted chunks, too.

Anyway, for me it seems it was actually the
 
```
borg compact --cleanup-commits ::
```

that shaved off ~800 GB off my 1500 GB disk usage. Wow.

(Running `borg --progress --verbose compact ::` afterwards then only `freed about 1.76 MB repository space`.)

**Comment by ThomasWaldmann** [comment]
Guess you forgot to read the changelog, which is strongly recommended when doing a bigger version jump:

https://borgbackup.readthedocs.io/en/stable/changes.html#change-log

There, under Compatibility Notes:

**freeing repository space only happens when “borg compact” is invoked.**

**Comment by codethief** [comment]
@ThomasWaldmann I apologize if my comment came across the wrong way. I understand that you are documenting everything diligently and I'm grateful for that.

However, I would have never expected such a breaking change in a new minor version, which is why it only occurred to me very late during my debugging that something might have changed in Borg. 

In any case, what do you think about my suggestion to announce the most crucial changelogs / breaking changes in the CLI?

**Comment by ThomasWaldmann** [comment]
We try to follow semantic versioning.

If the last digit of the version number changes (e.g. 1.2.3 to 1.2.4), expect mostly bug fixes and no big changes. Reading the changelog might be interesting, but is usually not required. Most of borg releases are of this kind.

If the middle digit changes (like 1.1.x to 1.2.x), there can be also bigger changes, but nothing will fundamentally break. In such cases, always read the changelog. Happens every few years usually.

If the first digit changes (like 1.x.x to 2.x.x), there can be fundamentally breaking changes. Like that upcoming borg2 will not "just work" with your existing borg 1.x repos. Reading changelog and upgrade guide is mandatory. First time this will happen is at borg2 release (and we try to get all breaking changes into that release to hopefully not require more breaking releases anytime soon after that). borg 1.2.x and borg 2 will be maintained in parallel for a while, so people will have time to adapt.

For the case of `borg compact`, a version change only at the hoster / "borg serve" side (1.1.x to 1.2.x) does not change behaviour (IIRC). But if you upgrade client-side, you'll have to run `borg compact` to free space.

BTW, just as a general hint, some hosters have multiple borg versions installed and you can choose using `--remote-path=borg-x.y` or so. So you might have some control over that.

I would like to keep the change docs at one place (changelog) and not also have "change hint" management in the code, especially not with additional state / confirmation management you suggested.

BTW, in some cases, borg **will** warn you ahead - e.g. if we change or remove some cli option (e.g. in borg 1.2), there usually will be a deprecation warning (e.g. in 1.1.latest) so that you'll notice it if you are using something that will change / go away.
Depending on the packaging policy (e.g. of some stable linux dists), you might never see some deprecation warnings though, because such dists do not upgrade to newer patch releases over time, but stick to the original version and just patch security and other major issues.
